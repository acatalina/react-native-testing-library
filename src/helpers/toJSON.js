// @flow
import { ReactTestInstance } from 'react-test-renderer';

type ReactTestInstanceJSON = {
  type: string,
  props: { [propName: string]: any },
  children: null | Array<ReactTestInstanceJSON>,
  $$typeof?: Symbol,
};

type ToJSONOptions = {
  omitProps: Array<string>,
};

export type ToJSON = (
  instance: ReactTestInstance,
  options?: ToJSONOptions
) => ReactTestInstanceJSON;

const toJSON: ToJSON = (
  { _fiber: { stateNode = null } = {} } = {},
  options = {}
) => {
  if (!stateNode) {
    return null;
  } else if (
    stateNode.rootContainerInstance &&
    stateNode.rootContainerInstance.children.length === 0
  ) {
    return null;
  }

  return _toJSON(stateNode, options);
};

const _toJSON: ToJSON = (inst, { omitProps = [] }) => {
  if (inst.isHidden) {
    // Omit timed out children from output entirely. This seems like the least
    // surprising behavior. We could perhaps add a separate API that includes
    // them, if it turns out people need it.
    return null;
  }
  switch (inst.tag) {
    case 'TEXT':
      return inst.text;
    case 'INSTANCE': {
      // We don't include the `children` prop in JSON.
      // Instead, we will include the actual rendered children.
      // eslint-disable-next-line no-unused-vars
      const { children, ...props } = inst.props;

      // Convert all children to the JSON format
      const renderedChildren = inst.children.map((child) => {
        return _toJSON(child, { omitProps });
      });

      // Function props get noisy in debug output, so we'll exclude them
      // Also exclude any props configured via options.omitProps
      let renderedProps = {};
      Object.keys(props).filter((name) => {
        if (typeof props[name] !== 'function' && !omitProps.includes(name)) {
          renderedProps[name] = props[name];
        }
      });

      const json = {
        type: inst.type,
        props: renderedProps,
        children: renderedChildren,
      };

      Object.defineProperty(json, '$$typeof', {
        value: Symbol.for('react.test.json'),
      });
      return json;
    }
    default:
      throw new Error(`Unexpected node type in toJSON: ${inst.tag}`);
  }
};

export { toJSON };
