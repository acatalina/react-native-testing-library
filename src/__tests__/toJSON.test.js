import React from 'react';
import { Text, View } from 'react-native';
import { render } from '..';

import { toJSON } from '../helpers/toJSON';

test('it converts to json', () => {
  const testID = 'container';
  const { getByTestId } = render(
    <View>
      <View testID={testID}>
        <View>
          <Text>hello</Text>
        </View>
        <View style={{ flex: 1 }}>
          <Text onPress={jest.fn()}>world</Text>
          <Text>foo bar</Text>
        </View>
        <View />
      </View>
    </View>
  );

  expect(toJSON(getByTestId(testID))).toMatchSnapshot();
  expect(
    toJSON(getByTestId(testID), { omitProps: ['style'] })
  ).toMatchSnapshot();
});

test('it handles no arguments', () => {
  expect(toJSON()).toBeNull();
});

test('it handles hidden nodes', () => {
  expect(toJSON({ _fiber: { stateNode: { isHidden: true } } })).toBeNull();
});

test('it handles invalid nodes', () => {
  expect(() => toJSON({ _fiber: { stateNode: { tag: 'FAKE' } } })).toThrow();
});
