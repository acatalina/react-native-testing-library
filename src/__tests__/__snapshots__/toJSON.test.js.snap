// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it converts to json 1`] = `
<View
  testID="container"
>
  <View>
    <Text>
      hello
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text>
      world
    </Text>
    <Text>
      foo bar
    </Text>
  </View>
  <View />
</View>
`;

exports[`it converts to json 2`] = `
<View
  testID="container"
>
  <View>
    <Text>
      hello
    </Text>
  </View>
  <View>
    <Text>
      world
    </Text>
    <Text>
      foo bar
    </Text>
  </View>
  <View />
</View>
`;
